user_guide_src/source/database/query_builder/092.php

```diff
@@ -2,28 +2,53 @@
 
 $data = [
     [
-        'title' => 'Title 1',
-        'name'  => 'Name 1',
-        'date'  => 'Date 1',
+        'title'  => 'Title 1',
+        'author' => 'Author 1',
+        'name'   => 'Name 1',
+        'date'   => 'Date 1',
     ],
     [
-        'title' => 'Title 2',
-        'name'  => 'Name 2',
-        'date'  => 'Date 2',
+        'title'  => 'Title 2',
+        'author' => 'Author 2',
+        'name'   => 'Name 2',
+        'date'   => 'Date 2',
     ],
 ];
 
-$builder->updateBatch($data, 'title');
+$builder->updateBatch($data, ['title', 'author']);
+
+// OR
+$builder->setData($data)->onConstraint('title, author')->updateBatch();
+
+// OR
+$builder->setData($data)
+    ->setAlias('u')
+    ->onConstraint(['`mytable`.`title`' => '`u`.`title`', 'author' => new RawSql('`u`.`author`')])
+    ->updateBatch();
+
+// OR
+$builder->setData($data)
+    ->setAlias('u')
+    ->onConstraint(new RawSql('`mytable`.`title` = `u`.`title` AND `mytable`.`author` = `u`.`author`'))
+    ->updateFields(['last_update' => new RawSql('CURRENT_TIMESTAMP()')], true)
+    ->updateBatch();
+
+// OR
+foreach ($data as $row) {
+    $builder->setData($row);
+}
+$builder->onConstraint('title, author')->updateBatch();
```

Comment:
All of the examples all generate the same SQL (except one additional item in one). The idea here was to demonstrate all the different ways one might construct a statement. There is still come features I want to add to this.. the use of join() and where(). When I add those I thought I might make some more examples. I though we might want to make a separate page though at some point because it could get quite lengthy. Until then I thought it would suffice to demonstrate the basic methods available. Its not suppose to be an advanced lesson in SQL but just show what is available. A second page might could show some practical examples of using advanced methods.

Reply:
Okay, thank you for pointing this out. I misled myself by looking at the `last_update` part in the QB, which is present in the produced query. Now I even see there is a comment in the produced SQL... 😓 

Can we sort these examples so the comment becomes accurate? Now, the last scenario doesn't have the `last_update` part in the QB.

Yes, I think more advanced examples that use these new methods (including `RawSql`) would be desirable. A separate page would be great.

Reply:
Ya I added another example and forgot to reorder them

Reply:
I'd like to submit some more PRs though before adding a special page of examples. I'd like to add upsert as well as some more features to insert and update.. such as joining tables, insert/update from another table or a query. I'd just like to have a full feature set and then decide how best to demonstrate everything. I definitely will add more later.. these changes will generate endless possibilities.